bundles/org.openhab.core.addon.marketplace.karaf/src/main/java/org/openhab/core/addon/marketplace/karaf/internal/community/CommunityKarafAddonHandler.java

```diff
@@ -0,0 +1,189 @@
+    private static final String KAR_DOWNLOAD_URL_PROPERTY = "kar_download_url";
+
+    private final Logger logger = LoggerFactory.getLogger(CommunityKarafAddonHandler.class);
+
+    private final FeaturesService featuresService;
+
+    @Activate
+    public CommunityKarafAddonHandler(@Reference FeaturesService featuresService) {
+        this.featuresService = featuresService;
+        ensureCachedBundlesAreInstalled();
+    }
+
+    @Override
+    public boolean supports(String type, String contentType) {
+        // we support only certain extension types, and only as pure OSGi bundles
+        return SUPPORTED_EXT_TYPES.contains(type) && KAR_CONTENT_TYPE.equals(contentType);
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    public boolean isInstalled(String addonId) {
+        try {
+            File addonDirectory = getAddonCacheDirectory(addonId);
+            List<URI> repositories = Arrays.stream(featuresService.listRepositories()).map(Repository::getURI)
+                    .collect(Collectors.toList());
+            if (addonDirectory.exists() && addonDirectory.isDirectory()) {
+                return Files.list(addonDirectory.toPath()).filter(path -> path.endsWith(".kar")).findFirst()
+                        .map(Path::toUri).map(repositories::contains).orElse(false);
+            }
+        } catch (Exception e) {
+            logger.warn("Failed to determine installation status for {}: ", addonId, e);
+        }
+
+        return false;
+    }
+
+    @Override
+    public void install(Addon addon) throws MarketplaceHandlerException {
+        URL sourceUrl;
+        try {
+            sourceUrl = new URL((String) addon.getProperties().get(KAR_DOWNLOAD_URL_PROPERTY));
+            addBundleToCache(addon.getId(), sourceUrl);
+            installFromCache(addon.getId());
+        } catch (MalformedURLException e) {
+            throw new MarketplaceHandlerException("Malformed source URL: " + e.getMessage());
+        }
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    public void uninstall(Addon addon) throws MarketplaceHandlerException {
+        try {
+            File addonPath = getAddonCacheDirectory(addon.getId());
+            List<URI> repositories = Arrays
+                    .stream(Objects.requireNonNullElse(featuresService.listRepositories(), new Repository[] {}))
+                    .map(Repository::getURI).collect(Collectors.toList());
+            if (addonPath.exists() && addonPath.isDirectory()) {
+                for (File file : Objects.requireNonNullElse(addonPath.listFiles(), new File[] {})) {
+                    if (repositories.contains(file.toURI())) {
+                        featuresService.removeRepository(file.toURI(), true);
+                    }
+                    file.delete();
+                }
+            }
+            addonPath.delete();
+        } catch (Exception e) {
+            throw new MarketplaceHandlerException("Failed uninstalling bundle: " + e.getMessage());
```

Comment:
Is it possible to add the stacktrace to all the `MarketplaceHandlerException`s? It would simplify debugging.

Reply:
Agreed, I'll extend the `MarketplaceHandlerException` with a `Throwable cause`.